clients/drcachesim/reader/zipfile_file_reader.cpp

```diff
@@ -130,4 +131,100 @@ file_reader_t<zipfile_reader_t>::is_complete()
+                                                          OUT bool *eof)
+{
+    if (instruction_count == 0)
+        return true;
+    VPRINT(this, 2, "Thread #%zd skipping %" PRIi64 " instrs\n", thread_index,
+           instruction_count);
+    trace_entry_t timestamp = {};
+    trace_entry_t cpu = {};
+    const memref_t &memref = **this;
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP) {
+        timestamp = entry_copy_;
+    }
+    zipfile_reader_t *zipfile = &input_files_[thread_index];
+    // We assume our unzGoToNextFile loop is plenty performant and we don't need to
+    // know the chunk names to use with a single unzLocateFile.
+    uint64_t stop_count_ = cur_instr_count_ + instruction_count + 1;
+    VPRINT(this, 2,
+           "stop=%" PRIi64 " cur=%" PRIi64 " chunk=%" PRIi64 " est=%" PRIi64 "\n",
+           stop_count_, cur_instr_count_, chunk_instr_count_,
+           cur_instr_count_ +
+               (chunk_instr_count_ - (cur_instr_count_ % chunk_instr_count_)));
+    while (cur_instr_count_ +
+               (chunk_instr_count_ - (cur_instr_count_ % chunk_instr_count_)) <
+           stop_count_) {
+        if (unzCloseCurrentFile(zipfile->file) != UNZ_OK)
+            return false;
+        int res = unzGoToNextFile(zipfile->file);
+        if (res != UNZ_OK) {
+            if (res == UNZ_END_OF_LIST_OF_FILE) {
+                VPRINT(this, 2, "Thread #%zd hit EOF\n", thread_index);
+                *eof = true;
+            }
+            return false;
+        }
+        if (unzOpenCurrentFile(zipfile->file) != UNZ_OK)
+            return false;
+        cur_instr_count_ += chunk_instr_count_ - (cur_instr_count_ % chunk_instr_count_);
+        VPRINT(this, 2, "Thread #%zd at %" PRIi64 " instrs at start of new chunk\n",
+               thread_index, cur_instr_count_);
+        VPRINT(this, 2,
+               "zip chunk stop=%" PRIi64 " cur=%" PRIi64 " chunk=%" PRIi64 " est=%" PRIi64
+               "\n",
+               stop_count_, cur_instr_count_, chunk_instr_count_,
+               cur_instr_count_ +
+                   (chunk_instr_count_ - (cur_instr_count_ % chunk_instr_count_)));
+        // Clear cached data from the prior chunk.
+        zipfile->cur_buf = zipfile->max_buf;
+    }
+    // We have to linearly walk the last mile.
+    // We need to present a timestamp+cpu so we reset this field so process_input_entry()
+    // will not skip the first pair in this new chunk.
+    last_timestamp_instr_count_ = cur_instr_count_;
+    while (cur_instr_count_ < stop_count_) {
+        if (!read_next_thread_entry(thread_index, &entry_copy_, eof))
+            return false;
+        // We need to pass up memrefs for the final skipped instr, but we don't
+        // want to process_input_entry() on the first unskipped instr so we can
+        // insert the timestamp+cpu first.
+        if (cur_instr_count_ + 1 == stop_count_ &&
```

Comment:
So the loop condition is never false?

Reply:
Optionally, to have a meaningful ending while-loop condition, we could move read_next_thread_entry to the loop's end, and also have another one before the loop.

Reply:
Added a comment that the loop condition is not reached.